### ISSUE TITLE
xgboost GPU crash

### ISSUE THREAD

**Issue by szilard** [dmlc/xgboost#2442]
Training on AWS GPU on 10 million records crashes, while on 1M it works fine:

```
dtrain = xgb.DMatrix('train.libsvm')
params = {'max_depth':10, 'eta':0.1, 'objective':'binary:logistic', 'updater':'grow_gpu_hist'}
num_round = 100
md = xgb.train(params, dtrain, num_round)
```

full code [here](https://github.com/szilard/GBM-perf/blob/master/run/2a-xgboost-gpu.py)

GPU memory usage on 1M: 
```
[0] Tesla K80        | 61'C,  76 % |   154 / 11439 MB | python(152M)
```
so 10M data should fit on the GPU well.

Error I get:

```
[07:25:10] Tree method is automatically selected to be 'approx' for faster speed. to use old behavior(exact greedy algorithm on single machine), set tree_method to 'exact'
[07:25:16] Device: [0] Tesla K80
[07:25:16] /home/ubuntu/dwnl/xgboost/dmlc-core/include/dmlc/././logging.h:300: [07:25:16] /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/gpu_hist_builder.cu:582: Check failed: fmat.SingleColBlock() grow_gpu_hist: must have single column block. Try setting 'tree_method' parameter to 'exact'

Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3c) [0x7effa9f7654c]
[bt] (1) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree14GPUHistBuilder8InitDataERKSt6vectorINS_9bst_gpairESaIS3_EERNS_7DMatrixERKNS_7RegTreeE+0x948) [0x7effaa137ad8]
[bt] (2) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree14GPUHistBuilder6UpdateERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEPNS_7RegTreeE+0x11) [0x7effaa137f61]
[bt] (3) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree12GPUHistMakerINS0_9GradStatsEE6UpdateERKSt6vectorINS_9bst_gpairESaIS5_EEPNS_7DMatrixERKS4_IPNS_7RegTreeESaISD_EE+0x212) [0x7effaa0cad02]
[bt] (4) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree13BoostNewTreesERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEiPS2_ISt10unique_ptrINS_7RegTreeESt14default_deleteISB_EESaISE_EE+0x597) [0x7effa9ff4e37]
[bt] (5) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree7DoBoostEPNS_7DMatrixEPSt6vectorINS_9bst_gpairESaIS5_EEPNS_11ObjFunctionE+0x946) [0x7effa9ff5c76]
[bt] (6) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost11LearnerImpl13UpdateOneIterEiPNS_7DMatrixE+0x260) [0x7effaa0c5440]
[bt] (7) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(XGBoosterUpdateOneIter+0x2b) [0x7effa9f6d6bb]
[bt] (8) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call_unix64+0x4c) [0x7effcdf6057c]
[bt] (9) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call+0x1f5) [0x7effcdf5fcd5]

[07:25:16] /home/ubuntu/dwnl/xgboost/dmlc-core/include/dmlc/././logging.h:300: [07:25:16] /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/updater_gpu.cc:73: GPU plugin exception: [07:25:16] /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/gpu_hist_builder.cu:582: Check failed: fmat.SingleColBlock() grow_gpu_hist: must have single column block. Try setting 'tree_method' parameter to 'exact'

Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3c) [0x7effa9f7654c]
[bt] (1) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree14GPUHistBuilder8InitDataERKSt6vectorINS_9bst_gpairESaIS3_EERNS_7DMatrixERKNS_7RegTreeE+0x948) [0x7effaa137ad8]
[bt] (2) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree14GPUHistBuilder6UpdateERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEPNS_7RegTreeE+0x11) [0x7effaa137f61]
[bt] (3) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree12GPUHistMakerINS0_9GradStatsEE6UpdateERKSt6vectorINS_9bst_gpairESaIS5_EEPNS_7DMatrixERKS4_IPNS_7RegTreeESaISD_EE+0x212) [0x7effaa0cad02]
[bt] (4) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree13BoostNewTreesERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEiPS2_ISt10unique_ptrINS_7RegTreeESt14default_deleteISB_EESaISE_EE+0x597) [0x7effa9ff4e37]
[bt] (5) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree7DoBoostEPNS_7DMatrixEPSt6vectorINS_9bst_gpairESaIS5_EEPNS_11ObjFunctionE+0x946) [0x7effa9ff5c76]
[bt] (6) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost11LearnerImpl13UpdateOneIterEiPNS_7DMatrixE+0x260) [0x7effaa0c5440]
[bt] (7) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(XGBoosterUpdateOneIter+0x2b) [0x7effa9f6d6bb]
[bt] (8) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call_unix64+0x4c) [0x7effcdf6057c]
[bt] (9) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call+0x1f5) [0x7effcdf5fcd5]



Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3c) [0x7effa9f7654c]
[bt] (1) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree12GPUHistMakerINS0_9GradStatsEE6UpdateERKSt6vectorINS_9bst_gpairESaIS5_EEPNS_7DMatrixERKS4_IPNS_7RegTreeESaISD_EE+0x2dd) [0x7effaa0cadcd]
[bt] (2) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree13BoostNewTreesERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEiPS2_ISt10unique_ptrINS_7RegTreeESt14default_deleteISB_EESaISE_EE+0x597) [0x7effa9ff4e37]
[bt] (3) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree7DoBoostEPNS_7DMatrixEPSt6vectorINS_9bst_gpairESaIS5_EEPNS_11ObjFunctionE+0x946) [0x7effa9ff5c76]
[bt] (4) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost11LearnerImpl13UpdateOneIterEiPNS_7DMatrixE+0x260) [0x7effaa0c5440]
[bt] (5) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(XGBoosterUpdateOneIter+0x2b) [0x7effa9f6d6bb]
[bt] (6) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call_unix64+0x4c) [0x7effcdf6057c]
[bt] (7) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call+0x1f5) [0x7effcdf5fcd5]
[bt] (8) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(_ctypes_callproc+0x3e6) [0x7effcdf57376]
[bt] (9) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(+0x9db3) [0x7effcdf4edb3]

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/training.py", line 204, in train
    xgb_model=xgb_model, callbacks=callbacks)
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/training.py", line 74, in _train_internal
    bst.update(dtrain, i, obj)
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/core.py", line 819, in update
    _check_call(_LIB.XGBoosterUpdateOneIter(self.handle, iteration, dtrain.handle))
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/core.py", line 127, in _check_call
    raise XGBoostError(_LIB.XGBGetLastError())
xgboost.core.XGBoostError: [07:25:16] /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/updater_gpu.cc:73: GPU plugin exception: [07:25:16] /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/gpu_hist_builder.cu:582: Check failed: fmat.SingleColBlock() grow_gpu_hist: must have single column block. Try setting 'tree_method' parameter to 'exact'

Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3c) [0x7effa9f7654c]
[bt] (1) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree14GPUHistBuilder8InitDataERKSt6vectorINS_9bst_gpairESaIS3_EERNS_7DMatrixERKNS_7RegTreeE+0x948) [0x7effaa137ad8]
[bt] (2) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree14GPUHistBuilder6UpdateERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEPNS_7RegTreeE+0x11) [0x7effaa137f61]
[bt] (3) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree12GPUHistMakerINS0_9GradStatsEE6UpdateERKSt6vectorINS_9bst_gpairESaIS5_EEPNS_7DMatrixERKS4_IPNS_7RegTreeESaISD_EE+0x212) [0x7effaa0cad02]
[bt] (4) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree13BoostNewTreesERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEiPS2_ISt10unique_ptrINS_7RegTreeESt14default_deleteISB_EESaISE_EE+0x597) [0x7effa9ff4e37]
[bt] (5) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree7DoBoostEPNS_7DMatrixEPSt6vectorINS_9bst_gpairESaIS5_EEPNS_11ObjFunctionE+0x946) [0x7effa9ff5c76]
[bt] (6) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost11LearnerImpl13UpdateOneIterEiPNS_7DMatrixE+0x260) [0x7effaa0c5440]
[bt] (7) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(XGBoosterUpdateOneIter+0x2b) [0x7effa9f6d6bb]
[bt] (8) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call_unix64+0x4c) [0x7effcdf6057c]
[bt] (9) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call+0x1f5) [0x7effcdf5fcd5]



Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3c) [0x7effa9f7654c]
[bt] (1) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree12GPUHistMakerINS0_9GradStatsEE6UpdateERKSt6vectorINS_9bst_gpairESaIS5_EEPNS_7DMatrixERKS4_IPNS_7RegTreeESaISD_EE+0x2dd) [0x7effaa0cadcd]
[bt] (2) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree13BoostNewTreesERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEiPS2_ISt10unique_ptrINS_7RegTreeESt14default_deleteISB_EESaISE_EE+0x597) [0x7effa9ff4e37]
[bt] (3) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree7DoBoostEPNS_7DMatrixEPSt6vectorINS_9bst_gpairESaIS5_EEPNS_11ObjFunctionE+0x946) [0x7effa9ff5c76]
[bt] (4) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost11LearnerImpl13UpdateOneIterEiPNS_7DMatrixE+0x260) [0x7effaa0c5440]
[bt] (5) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(XGBoosterUpdateOneIter+0x2b) [0x7effa9f6d6bb]
[bt] (6) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call_unix64+0x4c) [0x7effcdf6057c]
[bt] (7) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call+0x1f5) [0x7effcdf5fcd5]
[bt] (8) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(_ctypes_callproc+0x3e6) [0x7effcdf57376]
[bt] (9) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(+0x9db3) [0x7effcdf4edb3]
```

System:
```
p2.xlarge   K80 GPU
Nvidia CUDA 8.0
Operating System: Ubuntu 16.04.2 LTS

Compiler: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0

If installing from source, please provide 
The commit hash (`git rev-parse HEAD`)
6776292951565c8cd72e69afd9d94de1474f00c0
```

**Comment by szilard** [comment]
also per @RAMitchell 's suggestion i tried to compile with `-DGPU_COMPUTE_VER="37"` but the behavior is the same

**Comment by szilard** [comment]
and the error is same on latest `1899f9e744c746118e8a44b4cbb72f76c0616535` xgboost

**Comment by RAMitchell** [comment]
Try setting tree method to exact to prevent xgboost from automatically switching to distributed mode.

**Comment by szilard** [comment]
You mean `'updater':'grow_gpu'` ? I tried that already and it crashed too.

**Comment by RAMitchell** [comment]
I am trying to reproduce your crash on a P2 instance.

It is unclear where the following two files come from:
dtrain = xgb.DMatrix('train.libsvm')
dtest = xgb.DMatrix('test.libsvm')

Can you link me to the script that generated them or upload to S3?

**Comment by szilard** [comment]
Sorry about that. I added them to S3 and the info [here](https://github.com/szilard/GBM-perf/blob/master/data.txt#L11-L18).

One thing I realized is that I did this primarily for LightGBM, which can deal with categorical variables using integer encoding. So using this for xgboost rather than 1-hot encoding might not be the "proper" corresponding way, but anyway.

**Comment by szilard** [comment]
Also, note that for each size (1M, 10M records) you need to use the corresponding test file (not the same test file as for CSV because the encoding).

**Comment by RAMitchell** [comment]
Setting parameters as follows works fine for me.
params = {'max_depth':10, 'eta':0.1, 'objective':'binary:logistic', 'updater':'grow_gpu', 'tree_method':'exact'}

I will work on a PR presently to add tree_method parameters for GPU algorithms to prevent this happening again.

I get 62s for grow_gpu_hist and 301s for grow_gpu.

I think your AUC is wrong for GPU XGBoost because it will be treating categorical variables as numerical.

Additionally I suggest increasing the number of boosting iterations to at least 500. This is because all of the gradient boosting algorithms that you are testing contain some level of fixed overhead and 100 iterations is a little small for a typical use case.

**Comment by RAMitchell** [comment]
Any update on this @szilard? I have changed things so you can use 'tree_method':'gpu_hist' or 'tree_method':'gpu_exact' without modifying the updater and also added compute flags by default for the k80.

**Comment by szilard** [comment]
Sorry @RAMitchell, I really appreciate your effort, but I did not get a chance to look at it yet. Also next week is useR, so probably I'll get to it the week after. But make no mistake, I will. Thanks.

**Comment by szilard** [comment]
Hi @RAMitchell, I installed latest xgboost `commit 64c8f6fa6d01e284143f6cdaab3ade163924ceca` with GPU support (apparently successfully) but running `params = {'max_depth':10, 'eta':0.1, 'objective':'binary:logistic', 'updater':'grow_gpu', 'tree_method':'exact'}` gets me now error even on the small (1M) dataset:

```
[12:21:00] /home/ubuntu/dwnl/xgboost/dmlc-core/include/dmlc/././logging.h:300: [12:21:00] /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/updater_gpu.cu:40: GPU plugin exception: /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/exact/../common.cuh(207): invalid device function


Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3c) [0x7f15657d440c]
[bt] (1) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree8GPUMaker6UpdateERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixERKS2_IPNS_7RegTreeESaISB_EE+0x4143) [0x7f15659be013]
[bt] (2) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree13BoostNewTreesERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEiPS2_ISt10unique_ptrINS_7RegTreeESt14default_deleteISB_EESaISE_EE+0x8c3) [0x7f1565878723]
[bt] (3) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree7DoBoostEPNS_7DMatrixEPSt6vectorINS_9bst_gpairESaIS5_EEPNS_11ObjFunctionE+0xa60) [0x7f15658799e0]
[bt] (4) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost11LearnerImpl13UpdateOneIterEiPNS_7DMatrixE+0x22b) [0x7f1565888dab]
[bt] (5) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(XGBoosterUpdateOneIter+0x27) [0x7f15657c6047]
[bt] (6) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call_unix64+0x4c) [0x7f1589f9f57c]
[bt] (7) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call+0x1f5) [0x7f1589f9ecd5]
[bt] (8) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(_ctypes_callproc+0x3e6) [0x7f1589f96376]
[bt] (9) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(+0x9db3) [0x7f1589f8ddb3]

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/training.py", line 204, in train
    xgb_model=xgb_model, callbacks=callbacks)
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/training.py", line 74, in _train_internal
    bst.update(dtrain, i, obj)
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/core.py", line 827, in update
    dtrain.handle))
  File "/home/ubuntu/dwnl/xgboost/python-package/xgboost/core.py", line 130, in _check_call
    raise XGBoostError(_LIB.XGBGetLastError())
xgboost.core.XGBoostError: [12:21:00] /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/updater_gpu.cu:40: GPU plugin exception: /home/ubuntu/dwnl/xgboost/plugin/updater_gpu/src/exact/../common.cuh(207): invalid device function


Stack trace returned 10 entries:
[bt] (0) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN4dmlc15LogMessageFatalD1Ev+0x3c) [0x7f15657d440c]
[bt] (1) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost4tree8GPUMaker6UpdateERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixERKS2_IPNS_7RegTreeESaISB_EE+0x4143) [0x7f15659be013]
[bt] (2) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree13BoostNewTreesERKSt6vectorINS_9bst_gpairESaIS3_EEPNS_7DMatrixEiPS2_ISt10unique_ptrINS_7RegTreeESt14default_deleteISB_EESaISE_EE+0x8c3) [0x7f1565878723]
[bt] (3) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost3gbm6GBTree7DoBoostEPNS_7DMatrixEPSt6vectorINS_9bst_gpairESaIS5_EEPNS_11ObjFunctionE+0xa60) [0x7f15658799e0]
[bt] (4) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(_ZN7xgboost11LearnerImpl13UpdateOneIterEiPNS_7DMatrixE+0x22b) [0x7f1565888dab]
[bt] (5) /home/ubuntu/dwnl/xgboost/python-package/xgboost/../../lib/libxgboost.so(XGBoosterUpdateOneIter+0x27) [0x7f15657c6047]
[bt] (6) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call_unix64+0x4c) [0x7f1589f9f57c]
[bt] (7) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(ffi_call+0x1f5) [0x7f1589f9ecd5]
[bt] (8) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(_ctypes_callproc+0x3e6) [0x7f1589f96376]
[bt] (9) /home/ubuntu/anaconda2/lib/python2.7/lib-dynload/_ctypes.so(+0x9db3) [0x7f1589f8ddb3]
```

Any ideas why?

**Comment by RAMitchell** [comment]
This has uncovered a bug in my latest cmake changes. Building with 'cmake .. -DPLUGIN_UPDATER_GPU=ON -DGPU_COMPUTE_VER="37"' will fix your problem. I will patch this shortly.

Note that you can set the tree method directly now to 'gpu_exact'/'gpu_hist' and not set the updater.

**Comment by szilard** [comment]
Cool, thanks @RAMitchell , it works now, 6sec (1M) and 50sec (10M) (all with hist) with `params = {'max_depth':10, 'eta':0.1, 'objective':'binary:logistic', 'tree_method':'gpu_hist'}` 

I have updated my repo: https://github.com/szilard/GBM-perf/commit/ef0fb33bc48180a4b390f05d5f67c8cb835bbf9a

**Comment by szilard** [comment]
Also, the integer encoding I use in the svmlight files is not the proper way for xgboost (I did that for lightgbm since they implemented a special way to handle categorical variables). Would there one-hot encoding with sparse (svmlight) be the proper way for xgboost for comparison?

**Comment by RAMitchell** [comment]
Im not too familiar with LightGbm but yes sparse one hot encoding is the way to go for categoricals in Xgboost.

**Comment by szilard** [comment]
OK, I'll do that soon.

Any plans to include the GPU capability in the xgboost R package?

**Comment by RAMitchell** [comment]
It should be available through h2o R. Adding it to XGBoost will be a little more difficult but maybe in the next six months.

**Comment by szilard** [comment]
Yeah, I've seen that (h2o xgboost from R), though that involves a few extra components. Anyway, I'll check out in 6 months then.

**Comment by RAMitchell** [comment]
I also notice that the h2o accuracy is a bit off. Is this using categoricals as numeric also? From the run-time of the H2O GPU algorithm I also assume you are using the exact method. You should also be able to use the histogram GPU method through H2O.

**Comment by szilard** [comment]
That's the thing, h2o might do various things besides just calling xgboost (surely it does stuff with categorical variables when it reads the data from CSV, but even after that h2o might do some "pre-processing" of data or other data transformations etc).

**Comment by szilard** [comment]
I've been trying to run now the same on the new g3.4xlarge instances that came out more recently (Tesla M60). I got some errors, so I recompiled with newest version but on the 1M dataset the run time is 50sec. I recompiled the newest version and ran it on the old K80 as well and run time is 50sec there as well. Run time used to be 6sec, something has changed. Any ideas?

**Comment by RAMitchell** [comment]
I'll have a look in the next few days. Appreciate the feedback.

**Comment by szilard** [comment]
Cool, thanks.

**Comment by RAMitchell** [comment]
A week or two ago I changed the histogram aggregation to use double precision atomic operations. This works fine on Pascal GPUs but older Kepler/Maxwell architectures seem to perform significantly worse.

I have reverted to single precision atomic operations in the latest commit.

**Comment by szilard** [comment]
OK, thanks, I'll check again. I wish the main cloud providers would have decent GPUs, not the crap they've been offering (always years behind the curve). I friend of mine promised me to run these things on some physical GPUs, still waiting on him as well.

**Comment by yitang** [comment]
any updates?

**Comment by RAMitchell** [comment]
This is an older issue - please create a new issue if you are having a problem with the GPU version.
